src/app/Library/CrudPanel/Traits/FieldsProtectedMethods.php

```diff
@@ -103,28 +102,17 @@ protected function makeSureFieldHasRelationshipData($field)
             return $field;
         }
 
-        $extraFieldAttributes = $this->inferFieldAttributesFromRelationship($field['entity']);
+        $extraFieldAttributes = $this->inferFieldAttributesFromRelationship($field);
 
-        if ($extraFieldAttributes !== false) {
-            $field = array_merge($extraFieldAttributes, $field);
+        if (! empty($extraFieldAttributes)) {
+            $field = array_merge($field, $extraFieldAttributes);
```

Comment:
Hmm I see you've switched the parameter order here. This will overwrite what the user has already declared, which is a very unintuitive behaviour. Everywhere else in Backpack, what the developer defines is king. But here, if they define ```entity```, ```multiple```, ```pivot```, even ```type``` - Backpack will ignore their orders and set whatever Backpack thinks is best? I don't think this is a good way to proceed @pxpm .

Reply:
I think for _most attributes_ it's easy to assume if the developer typed something, that's what they want. They know what they're doing. So let's prefer their value instead of the one we guess.

The only attribute that is questionable though... is ```name```. That's why you did this, right? Because you want to overwrite ```product.article.title``` with ```product[article[title]]```? Am I right?

Reply:
Just did not change them all, I did the same for type: $relationship['type'] = $field['type'] ?? 'relationship';

So even if later we overwrite, the values that we get from relation will take into account user preferences, have a look at `type`.

 I already did that change: `$relationship['type'] = $field['type'] ?? 'relationship';`

I will do it for the other attributes, multiple pivot if you agree with this last solution.

Reply:
Ok, in that case I agree - if the user preference overwrites our guessing.  Let's do that please.

But what happens with ```name```? As far as I understand, with the current code it will get overwritten 100% of the times, when using the dot notation. Am I right? That's the only reason why you switched the order of array_merge parameters, right?

In this case, if ```name``` is the culprit. I'm wondering. What if we change the name inside the ```relationship.blade.php``` file? Is that not what we _actually_ need? The way I see it:
- Backpack (when adding or working with fields) knows about this field as ```product.category.name```
- Laravel (in validation) knows about this field as ```product.category.name```
- the only place where this needs to be ```product[category[name]]``` is inside HTML

Am I wrong on this? If we can do this, then it means we can switch back the order or array_merge parameters, and go back to the old paradigm: Backpack doesn't overwrite anything you wrote. It just tries to fill in the blanks.

What do you think about this @pxpm ? Am I missing something here?

Reply:
@tabacitu AFAIK if we make that change we will need to change all field files... because for example, if your model as a HasOne relation with address and that address has `line_1` db column, you set the field with `$this->crud->addField('address.name');` the included field file is not the relationship.blade.php, but the text.blade.php or it could be the datepicker if you did `'type' => 'datepicker'` or whatever.

Was this explanation good @tabacitu ?

Reply:
Damn it, you're right... Not overwriting the name _here_ probably means having to change ALL fields. And that's a big NO-NO.

**What if we split up ```makeSureFieldHasRelationshipData()``` into one method per attribute:**
- isolate the **overwriting** of the name into a separate method, ```overwriteFieldNameFromDotNotationToArray()```; that way, just by looking at its name you'll know it does something different than all other ```makeSureFieldHasX()``` methods; this method would overwrite the ```name```, but only in this particular case - when the name was using dot notation; am I right?
- create a new method for each attribute that gets guessed by the ```makeSureFieldHasNecessaryAttributes()``` method; then that method would probably look like something this:

```diff
    protected function makeSureFieldHasNecessaryAttributes($field)
    {
        $field = $this->makeSureFieldHasName($field);
+       $field = $this->overwriteFieldNameFromDotNotationToArray($field);
        $field = $this->makeSureFieldHasLabel($field);
        $field = $this->makeSureFieldHasEntity($field);
+       if (isset($field['entity'])) {
+           $field = $this->makeSureFieldHasAttribute($field);
+           $field = $this->makeSureFieldHasForeignKey($field);
+           $field = $this->makeSureFieldHasMultiple($field);
+           $field = $this->makeSureFieldHasPivot($field);
+       } 
-       $field = $this->makeSureFieldHasAttribute($field);
-       $field = $this->makeSureFieldHasRelationshipAttributes($field);
        $field = $this->makeSureFieldHasType($field);

        return $field;
    }
```
- remove ```getFieldAttributesFromRelationship()``` - each attribute it used to treat now has its own method;
- remove ```inferFieldAttributesFromRelationship()``` - it didn't do much, other than call the method above;

PROs:
- all logic for one field attribute is incapsulated in one method; if you don't like how an attribute was guessed, you know exactly where to look;
- each ```makeSure()``` method will be easy to read, since it will be just a series of if/switch statements;
- nothing gets overwritten; except when it does 😄 (name in case of dot notation); but then the method name makes it clear that it overwrites;
- much easier to create unit tests for, since we can test each attribute guessing individually;

CONs:
- ```makeSure()``` methods will probably me bigger overall than one method that has all of them;
- ```makeSure()``` methods will probably have multiple similar if/switch statements, so we repeat ourselves a bit; 
- ```makeSure()``` methods will _might_ have to be run in a certain order, if some of them depend on a different attribute being guessed earlier;

There's a tradeoff here to be made, between single-responsibility-principle and do-not-repeat-yourself. And I think in this case it would be better to choose single-responsibility. What do you think @pxpm - wouldn't this be cleaner - one method per attribute? It would definitely be less scarry to merge 😆 

PS. After we do that, and everything's working fine, I would probably rename all makeSure methods, from ```makeSureFieldHasPivot()``` to ```guessFieldPivot()```. I think that's shorter and more accurate. And the name also kind of makes you understand that if a Pivot isn't necessary - it won't get added. Then in the end the main method would probably look like this:
```php
    protected function makeSureFieldHasNecessaryAttributes($field)
    {
        $field = $this->guessFieldName($field);
        $field = $this->overwriteFieldNameFromDotNotationToArray($field);
        $field = $this->guessFieldLabel($field);
        $field = $this->guessFieldEntity($field);

        if ( isset($field['entity']) ) {
            $field = $this->guesFieldAttribute($field);
            $field = $this->guessFieldForeignKey($field);
            $field = $this->guessFieldMultiple($field);
            $field = $this->guessFieldPivot($field);
        } 

        $field = $this->guessFieldType($field);

        return $field;
    }
```

Reply:
I did the change .. it is way better than before.

The only thing that I dislike is the fact that we call 2 times `getRelationInstance()`, one for model, and another for field type.

Reply:
Much better. Love it.